Remove `cmake`

Jasper-Bekkers opened the pull request:
# Why?
Currently because of the way this crate is set up (calling `cmake`) it requires that any user of this crate has `cmake` on their system installed. We think this badly impacts developer experience when using `zmq` (or our own downstream crates). As an example of where this shows up (not just our Traverse Research crates), is for example in Google's [evcxr](https://github.com/google/evcxr/blob/main/evcxr_jupyter/README.md#windows) setup instructions for all platforms.

# What?

It's become pretty standard in the Rust ecosystem to avoid `cmake` and instead to mirror it's buildscripts directly in Rust, usually it's relatively straight forward (just move over some defines and point it to some cpp files).

This switches over to the `cc` crate so that people using this crate don't need `cmake` to be installed on their system, but instead can just use `cargo build` without any additional requirements. The setup for this PR mirrors for example the setup used in [breakpad-sys](https://github.com/EmbarkStudios/sentry-contrib-rust/blob/main/breakpad-sys/build.rs) by Embark Studios.

As a result of switching to `cc` directly - we can enable parallel builds, and speed up the build times for `zmq` quite a bit when using the `vendored` feature - potentially making the non-vendored use-case of this crate obsolete.

# Platform support
 - [x] Windows
 - [x] Linux
 - [x] macOS

jean-airoldie commented:
Hi, removing cmake dep is of course a good idea. From what i recall, I only used the cmake build because building zmq without it was god awfull. If you can make this to work with only `cc` as a dependency, I'll happily merge this.

However the are a couple things i would like to see before I can merge:
* Add a linux CI build (looks travis doesn't provide free OSS credit, so anything else will do, for instance appveyor seems to work find).
* Provide the same functionality, unless you show that said functionality is useless or deprecated.
* If you add support for other platforms, please add CI builds for these.

jean-airoldie commented on src/lib.rs:
```diff
@@ -213,108 +206,251 @@ impl Build {
+            // TODO: Write these to a neat directory inside `out`
+            let out_includes = PathBuf::from(std::env::var("OUT_DIR").unwrap());
+            // Write out an empty platform file: defines will be set through cc directly
+            std::fs::write(out_includes.join("platform.hpp"), "").unwrap();
```
Don't forget to flush after writing, otherwise this can create a race on some systems.

Jasper-Bekkers replied:
Done.

Jasper-Bekkers commented:
I'll switch the entire CI step to run on GitHub actions then if you don't mind, and work on getting more or less feature parity with what was there before.

How attached are you to breaking/non-breaking changes to this crate? Right now I've kept the API mostly the same but there's quite a few things that aren't being used upstream, or not needed anymore (for example the profile always seems to get set to the current building profile - however `cc` also does that so it'd be pretty easy to override though Cargo.toml instead).

Jasper-Bekkers commented:
The CI run on the fork has turned green now https://github.com/Traverse-Research/zeromq-src-rs/runs/5069154901?check_suite_focus=true

jean-airoldie commented:
> How attached are you to breaking/non-breaking changes to this crate?

I think it probably makes sense to do a breaking change since the CI is this crate is probably a lil lacking and I wouldn't want to run the risk of breaking someone's build because of the move to `cc`. Some platforms a pretty finicky to build zmq on so i think its safer this way.

> The CI run on the fork has turned green now https://github.com/Traverse-Research/zeromq-src-rs/runs/5069154901?check_suite_focus=true

Cool

jean-airoldie commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,107 @@
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
+        rust: [stable]
+        tests: ["", "--release", "--features=static"]
```
I think you probably want to test with libsodium also. The default encryption lib has terrible performance and the devs have said it shoudn't be used for more than testing. Can't remember where I saw that exactly but I tested it and performance was indeed terrible. So as soon as someone want's to use ZMQ in a public environment, it's gonna be needed. Whether I would trust ZMQ in a public setting is another issue however.

Jasper-Bekkers replied:
Thanks I'll add it to the test matrix. 

You'll see some commits fly by while I try to get mingw to work as well but it's a bit of a nightmare it would seem ;-)

Any platforms you care about that should be covered?

Jasper-Bekkers replied:
If tweetnacl isn't supposed to be used - maybe we should just make libsodium the default instead?

jean-airoldie commented:
As you [mentionned](https://github.com/jean-airoldie/zeromq-src-rs/pull/17#discussion_r799757407) i think it would make sense to make libsodium the default encryption lib otherwise it's kind of a footgun. I think the ideal solution would be to have a `curve` feature that enables both the `ZMQ_CURVE` flag and links to libsodium.

I'm not using ZMQ anymore to be clear, but in my archive repo [I have an example of how to do the libsodium linking](https://github.com/jean-airoldie/libzmq-rs/tree/master/libzmq-sys).

jean-airoldie commented:
> Any platforms you care about that should be covered, I'm considering dropping BSD from this PR.

Only platforms I care about are windows and linux, so feel free to drop BSD.

Jasper-Bekkers commented:
> As you [mentionned](https://github.com/jean-airoldie/zeromq-src-rs/pull/17#discussion_r799757407) i think it would make sense to make libsodium the default encryption lib otherwise it's kind of a footgun. I think the ideal solution would be to have a `curve` feature that enables both the `ZMQ_CURVE` flag and links to libsodium.
> 
> I'm not using ZMQ anymore to be clear, but in my archived repo [I have an example of how to do the libsodium linking](https://github.com/jean-airoldie/libzmq-rs/blob/master/libzmq-sys/build.rs#L38).
> 
> edit: fixed the link

Looking at this now I see why you've set it up like that at the time - I think I'll have to come back on making libsodium a default feature, since it looks like the rust crate for it has been deprecated a while ago. I'll just add a test for it and leave the integration as-is.

Jasper-Bekkers commented:
@jean-airoldie I've started using this PR in production now on our end, and it works on all the platforms we care about. I'll update the PR description and title a bit to reflect that, so for all intents and purposes this can be considered as "done" on my end.

There's quite a trail of nasty commits that I could clean up if you're intending not to squash this PR, but otherwise, unless you have no other feedback, I'd be happy to send this off.

jean-airoldie commented:
Here's a sanity checking patch file for libsodium in the `testcrate`.
[0001-Improve-libsodium-test.txt](https://github.com/jean-airoldie/zeromq-src-rs/files/8027020/0001-Improve-libsodium-test.txt)
You can apply it via `git am <filename>`.

jean-airoldie commented:
> Looking at this now I see why you've set it up like that at the time - I think I'll have to come back on making libsodium a default feature, since it looks like the rust crate for it has been deprecated a while ago.

If you look at the patch I previously linked, you can see the [libsodium-sys-stable](https://docs.rs/libsodium-sys-stable/latest/libsodium_sys/index.html) crate that is maintained by the libsodium creator. It would be pretty simple automatically link to it via a feature, in the same way I did in the testcrate.

jean-airoldie reviewed (changes requested):
Seems like there are also some build issues:

https://github.com/Traverse-Research/zeromq-src-rs/runs/5085617997?check_suite_focus=true

https://github.com/Traverse-Research/zeromq-src-rs/runs/5085618099?check_suite_focus=true

Jasper-Bekkers commented on src/lib.rs:
```diff
@@ -213,108 +185,234 @@ impl Build {
+
+        build.define("ZMQ_USE_CV_IMPL_STL11", "1");
+        build.define("ZMQ_BUILD_DRAFT_API", "1");
+        build.define("ZMQ_STATIC", "1");
```
I did this mostly because the upstream `zmq-sys` crate just hardcodes it to `true` regardless. Figured that path wasn't taken anymore and could be removed.

jean-airoldie replied:
I think it makes sense to always hardcode to static and remove the static feature because we're vendoring the dependency anyway.

Jasper-Bekkers commented on src/lib.rs:
```diff
@@ -213,108 +185,234 @@ impl Build {
+        add_c_sources(&mut build, path.join("external/sha1"), &["sha1.c"]);
+
+        build.define("ZMQ_USE_CV_IMPL_STL11", "1");
+        build.define("ZMQ_BUILD_DRAFT_API", "1");
```
Good point!

Jasper-Bekkers commented on src/lib.rs:
```diff
@@ -31,42 +62,14 @@ where
 }
 
 #[derive(Debug, Clone)]
-pub struct Artifacts {
```
Upstream just unconditionally called `print_metadata` regardless - figured I'd just do it directly. Can clean it up if it's still required, but it felt a but superfluous.

jean-airoldie replied:
If i remember correctly, i think the rationale behind print_metadata was so that the user can output it himself if he has a weird use case. It kinda feels like a non existent edge case, so I'm ok with removing this.

Jasper-Bekkers commented on src/lib.rs:
```diff
@@ -213,108 +185,234 @@ impl Build {
-        }
-
-        if self.perf_tool {
-            config.define("WITH_PERF_TOOL", "ON");
```
Thanks yeah it seems to be some kind of loose binary anyway, so I figured I'd remove it.

jean-airoldie commented on src/lib.rs:
```diff
@@ -213,108 +185,234 @@ impl Build {
 
-        if target.contains("msvc") && link_type == LinkType::Dynamic {
-            panic!("dynamic compilation is currently not supported on windows");
+            if target.contains("msvc") {
```
On second thoughts, I think we should always link libsodium statically but package it with the lib by default.

Jasper-Bekkers commented on src/lib.rs:
```diff
@@ -213,108 +185,234 @@ impl Build {
+            .cpp(true)
+            .define("ZMQ_BUILD_TESTS", "OFF")
+            .include(path.join("include"))
+            //.include(path.join("external/include/sha1"))
```
A small - test, removed.

jean-airoldie commented:
Like mentioned previously, I'm fine with going the full static linking route everywhere as long we also vendor libsodium behind a `curve` feature and we remove the `static` feature.

jean-airoldie commented:
Also update the `.gitmodule` file to the proper `libzmq` branch. I'm guessing it should be `v4.3.2`.

jean-airoldie commented:
Also checkout the `vendor` repo to the `v4.3.4` branch.

Jasper-Bekkers commented:
> Like mentioned previously, I'm fine with going the full static linking route everywhere as long we also vendor libsodium behind a `curve` feature and we remove the `static` feature.

Right now it's set up such that if libsodium is enabled, we automatically enable the curve feature. Would you want it the other way around? Or just to have one feature for both?

jean-airoldie commented:
> Right now it's set up such that if libsodium is enabled, we automatically enable the curve feature. Would you want it the other way around? Or just to have one feature for both?

It's better to have one feature. I think its better to use the CURVE feature because most users don't really care about implementation details of what exact encryption library is used.

Jasper-Bekkers commented:
> It's better to have one feature. I think its better to use the CURVE feature because most users don't really care about implementation details of what exact encryption library is used.

I see what you mean - the reason it is the way it is, is because the libsodium feature requires some additional info to know where to locate it and it's headers. Want me to switch it?

jean-airoldie commented:
> I see what you mean - the reason it is the way it is, is because the libsodium feature requires some additional info to know where to locate it and it's headers. Want me to switch it - we'll have to move the libsodium params to the curve feature which also doesn't quite sound ideal?

What I was thinking was to depend on `libsodium-sys-stable` and enable it via the curve future. This way we also vendor `libsodium`. I can do that part on another PR because the release if you want me to.

jean-airoldie commented on testcrate/src/lib.rs:
```diff
@@ -20,14 +25,27 @@ pub fn version() -> (i32, i32, i32) {
+    fn sodium_version_works() {
+        let version = sodium_version();
+        println!("{:?}", version.to_str().unwrap());
+        assert_eq!(version.to_str().unwrap(), "1.0.18");
```
Just thought about this, this is gonna break because of semver when the libsodium version increases. It would be better instead to: `assert!(version.to_str().unwrap().starts_with("1.0"));`

Jasper-Bekkers commented on src/lib.rs:
```diff
@@ -31,42 +62,14 @@ where
+        if target.contains("windows") {
+            println!("cargo:rustc-link-lib=dylib=iphlpapi");
         }
-        println!("cargo:include={}", self.include_dir.display());
```
I'm not sure what `cargo:include` is supposed to do here - I can't find any docs on it?

Jasper-Bekkers commented:
@jean-airoldie I /think/ I've addressed most of your feedback - thanks, and I've made sure the build is green. I'm calling it quits for tonight, please let me know if there is any more feedback and I'll address it - during the week.

Jasper-Bekkers commented:
@jean-airoldie We've been running with this for a few days now - do you think it's mergable and if so could you also spin a new release?

jean-airoldie commented:
Sorry, I totally forgot about this, i'll do this today.

jean-airoldie commented:
I'm thinking I'm gonna merge, then do some of the changes I wanted to do myself, so its faster. In this case I was thinking of removing the artifact structure entirely and link against libsodium automatically when curve is enabled.

jean-airoldie commented:
Thanks for your work! I'm sure making this build work was painful.
